src/Illuminate/Database/Query/Grammars/PostgresGrammar.php

```diff
@@ -242,4 +255,23 @@ public function compileTruncate(Builder $query)
     {
         return ['truncate '.$this->wrapTable($query->from).' restart identity' => []];
     }
+
+    /**
+     * Concatenate an array of segments, removing empties.
+     *
+     * @param  array   $segments
+     * @return string
+     */
+    protected function concatenate($segments)
+    {
+        if (! empty($segments['unions']) && (! empty($segments['orders']) || ! empty($segments['limit']))) {
+            array_unshift($segments, '(');
+            $unionIndex = array_search('unions', array_keys($segments), true);
+            array_splice($segments, $unionIndex, 0, ')');
+        }
+
+        return implode(' ', array_filter($segments, function ($value) {
+            return (string) $value !== '';
```

Comment:
cast applies to $value only. Return block copied from the base class unchanged. Made a change to use parent impl.

Only new code is the `if {..}` block, which adds ( as first element and ) before union, wrapping all expression.